File: PHP/One/Transforms/InferTypesPlugins/Helpers/InferTypesPlugin.php
<?php

namespace One\Transforms\InferTypesPlugins\Helpers\InferTypesPlugin;

use One\ErrorManager\ErrorManager;
use One\Ast\Expressions\Expression;
use One\Transforms\InferTypes\InferTypes;
use One\Ast\Statements\Statement;
use One\Ast\Types\Property;
use One\Ast\Types\Lambda;
use One\Ast\Types\Method;
use One\Ast\Types\IMethodBase;

class InferTypesPlugin {
    public $main;
    public $errorMan;
    public $name;
    
    function __construct($name) {
        $this->name = $name;
        $this->errorMan = null;
    }
    
    function canTransform($expr) {
        return false;
    }
    
    function canDetectType($expr) {
        return false;
    }
    
    function transform($expr) {
        return $expr;
    }
    
    function detectType($expr) {
        return false;
    }
    
    function handleProperty($prop) {
        return false;
    }
    
    function handleLambda($lambda) {
        return false;
    }
    
    function handleMethod($method) {
        return false;
    }
    
    function handleStatement($stmt) {
        return false;
    }
}

File: CSharp/One/Transforms/InferTypesPlugins/Helpers/InferTypesPlugin.cs
using One.Ast;
using One.Transforms;
using One;

namespace One.Transforms.InferTypesPlugins.Helpers
{
    public class InferTypesPlugin {
        public InferTypes main;
        public ErrorManager errorMan;
        public string name;
        
        public InferTypesPlugin(string name)
        {
            this.name = name;
            this.errorMan = null;
        }
        
        public virtual bool canTransform(Expression expr)
        {
            return false;
        }
        
        public virtual bool canDetectType(Expression expr)
        {
            return false;
        }
        
        public virtual Expression transform(Expression expr)
        {
            return expr;
        }
        
        public virtual bool detectType(Expression expr)
        {
            return false;
        }
        
        public virtual bool handleProperty(Property prop)
        {
            return false;
        }
        
        public virtual bool handleLambda(Lambda lambda)
        {
            return false;
        }
        
        public virtual bool handleMethod(IMethodBase method)
        {
            return false;
        }
        
        public virtual bool handleStatement(Statement stmt)
        {
            return false;
        }
    }
}